crates/chia-consensus/src/spendbundle_validation.rs

```diff
@@ -0,0 +1,400 @@
+use crate::consensus_constants::ConsensusConstants;
+use crate::gen::condition_tools::make_aggsig_final_message;
+use crate::gen::flags::{ALLOW_BACKREFS, DISALLOW_INFINITY_G1, ENABLE_MESSAGE_CONDITIONS};
+use crate::gen::opcodes::{
+    AGG_SIG_AMOUNT, AGG_SIG_ME, AGG_SIG_PARENT, AGG_SIG_PARENT_AMOUNT, AGG_SIG_PARENT_PUZZLE,
+    AGG_SIG_PUZZLE, AGG_SIG_PUZZLE_AMOUNT,
+};
+use crate::gen::owned_conditions::OwnedSpendBundleConditions;
+use crate::gen::validation_error::ErrorCode;
+use crate::spendbundle_conditions::get_conditions_from_spendbundle;
+use chia_bls::BlsCache;
+use chia_bls::PairingInfo;
+use chia_protocol::SpendBundle;
+use clvmr::{ENABLE_BLS_OPS_OUTSIDE_GUARD, ENABLE_FIXED_DIV};
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+
+// currently in mempool_manager.py
+// called in threads from pre_validate_spend_bundle()
+// returns (error, cached_results, new_cache_entries, duration)
+pub fn validate_clvm_and_signature(
+    spend_bundle: &SpendBundle,
+    max_cost: u64,
+    constants: &ConsensusConstants,
+    height: u32,
+    cache: &Arc<Mutex<BlsCache>>,
+) -> Result<(OwnedSpendBundleConditions, Vec<PairingInfo>, Duration), ErrorCode> {
+    let start_time = Instant::now();
+    let npcresult = get_conditions_from_spendbundle(spend_bundle, max_cost, height, constants)
+        .map_err(|e| e.1)?;
+    let iter = npcresult.spends.iter().flat_map(|spend| {
+        let condition_items_pairs = [
+            (AGG_SIG_PARENT, &spend.agg_sig_parent),
+            (AGG_SIG_PUZZLE, &spend.agg_sig_puzzle),
+            (AGG_SIG_AMOUNT, &spend.agg_sig_amount),
+            (AGG_SIG_PUZZLE_AMOUNT, &spend.agg_sig_puzzle_amount),
+            (AGG_SIG_PARENT_AMOUNT, &spend.agg_sig_parent_amount),
+            (AGG_SIG_PARENT_PUZZLE, &spend.agg_sig_parent_puzzle),
+            (AGG_SIG_ME, &spend.agg_sig_me),
+        ];
+        condition_items_pairs
+            .into_iter()
+            .flat_map(move |(condition, items)| {
+                let spend_clone = spend.clone();
+                items.iter().map(move |(pk, msg)| {
+                    (
+                        pk,
+                        make_aggsig_final_message(
+                            condition,
+                            msg.as_slice(),
+                            &spend_clone,
+                            constants,
+                        ),
+                    )
+                })
+            })
+    });
+    let unsafe_items = npcresult
+        .agg_sig_unsafe
+        .iter()
+        .map(|(pk, msg)| (pk, msg.as_slice().to_vec()));
+    let iter = iter.chain(unsafe_items);
+    // Verify aggregated signature
+    let (result, added) = cache
+        .lock()
+        .unwrap()
+        .aggregate_verify(iter, &spend_bundle.aggregated_signature);
```

Comment:
this function doesn't really need the cache. the python implementation just uses the cache as a way to get at the resulting pairings. I think we could disentangle this function from the cache entirely by just computing the pairings, validating them and returning them

Reply:
it's probably simpler to mimic the python implementation for now